internal/utils/gke.go

```diff
@@ -0,0 +1,508 @@
+	ns, id := Parse(config.Spec.CredentialContent)
+	secret, err := secretsCache.Get(ns, id)
+
+	if err != nil {
+		return nil, err
+	}
+
+	dataBytes := secret.Data["gkeCredentialConfig-data"]
+
+	data := string(dataBytes)
+
+	return GetServiceClient(ctx, data)
+}
+
+// ValidateCreateRequest checks a config for the ability to generate a create request
+func ValidateCreateRequest(config *gkev1.GKEClusterConfig) error {
+	//TODO: check if these can even be nill/empty in a inported cluster.
+	if config.Spec.ProjectID == "" {
+		return fmt.Errorf("project ID is required")
+	} else if config.Spec.Zone == "" && config.Spec.Region == "" {
+		return fmt.Errorf("zone or region is required")
+	} else if config.Spec.Zone != "" && config.Spec.Region != "" {
+		return fmt.Errorf("only one of zone or region must be specified")
+	} else if config.Spec.ClusterName == "" {
+		return fmt.Errorf("cluster name is required")
+	}
+
+	for _, np := range config.Spec.NodePools {
+		if np.Autoscaling.Enabled != nil && *np.Autoscaling.Enabled {
+			if np.Autoscaling.MaxNodeCount == nil || np.Autoscaling.MinNodeCount == nil {
+				return fmt.Errorf("min and maxNodeCount can't be nil")
+			}
+			if *np.Autoscaling.MinNodeCount < 1 || *np.Autoscaling.MaxNodeCount < *np.Autoscaling.MinNodeCount {
+				return fmt.Errorf("minNodeCount in the NodePool must be >= 1 and <= maxNodeCount")
+			}
+		}
+	}
+
+	if config.Spec.IPAllocationPolicy == nil {
+		return fmt.Errorf("IpAllocationPolicy can't be nil")
+	}
+
+	//check if cluster with same name exists
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	svc, err := GetServiceClient(ctx, config.Spec.CredentialContent)
+	if err != nil {
+		return err
+	}
+	operation, err := svc.Projects.Locations.Clusters.List(
+		LocationRRN(config.Spec.ProjectID, config.Spec.Region)).Context(ctx).Do()
+
+	for _, cluster := range operation.Clusters {
+		if cluster.Name == config.Spec.ClusterName {
+			return fmt.Errorf("cannot create cluster [%s] because a cluster in GKE exists with the same name", config.Spec.ClusterName)
+		}
+	}
+
+	/*
+		if !config.Spec.Imported {
+			cannotBeNilError := "field [%s] cannot be nil for non-import cluster [%s]"
+			if config.Spec.KubernetesVersion == nil {
+				return fmt.Errorf(cannotBeNilError, "kubernetesVersion", config.Name)
+			}
+			if config.Spec.SecretsEncryption == nil {
+				return fmt.Errorf(cannotBeNilError, "secretsEncryption", config.Name)
+			}
+			if config.Spec.Tags == nil {
+				return fmt.Errorf(cannotBeNilError, "tags", config.Name)
+			}
+			if config.Spec.Subnets == nil {
+				return fmt.Errorf(cannotBeNilError, "subnets", config.Name)
+			}
+			if config.Spec.SecurityGroups == nil {
+				return fmt.Errorf(cannotBeNilError, "securityGroups", config.Name)
+			}
+			if config.Spec.LoggingTypes == nil {
+				return fmt.Errorf(cannotBeNilError, "loggingTypes", config.Name)
+			}
+		}
+	*/
```

Comment:
Can this be removed? It looks like EKS stuff.

Reply:
Done